### PR TITLE
Add counter metrics for calculating state delta

### PR DESCRIPTION
**Pull request by erikjohnston**
This will allow us to measure how often we calculate state deltas in
event persistence that we would have been able to calculate at the same
time we calculated the state for the event.

### CODE REVIEW

File: synapse/storage/events.py [hunk 1]
```diff
@@ -52,6 +52,15 @@
 event_counter = metrics.register_counter(
     "persisted_events_sep", labels=["type", "origin_type", "origin_entity"]
 )
+state_delta_counter = metrics.register_counter(
```
Comment (richvdh) [inline synapse/storage/events.py]: please could you add comments to explain what these count?

File: synapse/storage/events.py [hunk 2]
```diff
@@ -389,6 +399,25 @@ def _persist_events(self, events_and_contexts, backfilled=False,
...
+                            if len(new_latest_event_ids) == 1:
+                                state_delta_single_event_counter.inc()
+
+                                # This is a fairly handwavey check to see if we could
+                                # have guessed what the delta would have been when
+                                # processing one of these events.
+                                # What we're interested in is if the latest extremities
+                                # were the same when we created the event as they are
+                                # now. We guess this by looking at the prev events and
+                                # checking if they match up, as when this server creates
```
Comment (richvdh) [inline synapse/storage/events.py]: what do you mean by "this server" ?

File: synapse/storage/events.py [hunk 3]
```diff
@@ -389,6 +399,25 @@ def _persist_events(self, events_and_contexts, backfilled=False,
...
+                                # a new event it will use the extremities as the prev
+                                # events.
+                                for ev, _ in ev_ctx_rm:
+                                    prev_event_ids = set(e for e, _ in ev.prev_events)
+                                    if latest_event_ids == prev_event_ids:
+                                        state_delta_reuse_delta_counter.inc()
+                                    break
```
Comment (richvdh) [inline synapse/storage/events.py]: we only look at one of the events in `ev_ctx_rm` ?

File: synapse/storage/events.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/storage/events.py]: As in, we are creating a new event locally, as opposed to coming in from federation

Comment (erikjohnston) [reply synapse/storage/events.py]: I could change it to say "when we create an event locally"? Not sure if that is less confusing or not.

File: synapse/storage/events.py [hunk 4]
```diff
@@ -389,6 +406,25 @@ def _persist_events(self, events_and_contexts, backfilled=False,
...
+                                # were the same when we created the event as they are
+                                # now. We guess this by looking at the prev events and
+                                # checking if they match up, as when this server creates
+                                # a new event it will use the extremities as the prev
+                                # events.
+                                for ev, _ in ev_ctx_rm:
+                                    prev_event_ids = set(e for e, _ in ev.prev_events)
+                                    if latest_event_ids == prev_event_ids:
+                                        state_delta_reuse_delta_counter.inc()
+                                        break
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: We look at each event until we find a match. If we were implementing this then if we found a match we'd just use the delta from that and return.

**Pull request merged**